// mappersmith/esm/test/enhanced-debugging.mjs
// src/test/enhanced-debugging.ts
import { sortedUrl, toSortedQueryString } from "../mocks/mock-utils.mjs";
import { createRequire } from "module";
var colors = {
  green: (s) => `\x1B[32m${s}\x1B[39m`,
  red: (s) => `\x1B[31m${s}\x1B[39m`,
  yellow: (s) => `\x1B[33m${s}\x1B[39m`,
  bgGreen: (s) => `\x1B[42m${s}\x1B[49m`,
  bgRed: (s) => `\x1B[41m${s}\x1B[49m`
};
var ttyTable = null;
var Diff = null;
var requireFromCwd = createRequire(process.cwd() + "/package.json");
try {
  ttyTable = requireFromCwd("tty-table");
} catch {
}
try {
  Diff = requireFromCwd("diff");
} catch {
}
var diffString = (stringOne, stringTwo) => {
  if (!Diff) {
    return `${colors.red(stringOne)} \u2192 ${colors.green(stringTwo)}`;
  }
  const diff = Diff.diffChars(stringOne, stringTwo);
  return diff.map(function(part) {
    return part.added ? colors.bgGreen(part.value) : part.removed ? colors.bgRed(part.value) : colors.green(part.value);
  }).join("");
};
var getMatchScore = (match) => {
  return (match.isExactMatch ? 10 : match.isPartialMatch ? 5 : 0) + (match.url.match ? 4 : 0) + (match.method.match ? 3 : 0) + (match.body.match ? 2 : 0) + (match.headers.match ? 1 : 0);
};
var sortMatches = (matches) => {
  return matches.sort((a, b) => getMatchScore(a) - getMatchScore(b));
};
var createTableForMocks = (matches) => {
  if (!ttyTable) {
    return sortMatches(matches).map((match) => {
      const indicator = match.isExactMatch ? colors.green("\u2713") : match.isPartialMatch ? colors.yellow("~") : colors.red("\u2717");
      const row = (label, { match: m, mockValue, requestValue }) => {
        const value = mockValue !== requestValue ? diffString(requestValue, mockValue) : mockValue;
        return `  ${bold(label.padEnd(8))} ${m ? colors.green("Yes") : colors.red("No ")}  ${m ? colors.green(value) : colors.red(value)}`;
      };
      return [
        `${indicator} ${bold(match.mockName)}`,
        row("Url", match.url),
        row("Method", match.method),
        row("Body", match.body),
        row("Headers", match.headers)
      ].join("\n");
    }).join("\n\n");
  }
  const tables = sortMatches(matches).map((match) => {
    const header = [
      {
        value: bold(`${match.mockName}`),
        width: 20,
        color: "white",
        headerColor: "white",
        align: "left"
      },
      { value: bold("Match"), width: 15, color: "white", headerColor: "white", align: "left" },
      { value: bold("Value/Diff"), width: 80, color: "white", headerColor: "white", align: "left" }
    ];
    const getColumnValuesFromMatch = ({ match: match2, mockValue, requestValue }) => {
      const value = mockValue !== requestValue ? diffString(requestValue, mockValue) : mockValue;
      return [
        match2 ? colors.green("Yes") : colors.red("No"),
        match2 ? colors.green(value) : colors.red(value)
      ];
    };
    const rows = [
      [bold("Url"), ...getColumnValuesFromMatch(match.url)],
      [bold("Method"), ...getColumnValuesFromMatch(match.method)],
      [bold("Body"), ...getColumnValuesFromMatch(match.body)],
      [bold("Headers"), ...getColumnValuesFromMatch(match.headers)]
    ];
    const table = ttyTable(header, rows, {
      borderStyle: "solid",
      borderColor: match.isExactMatch ? "green" : match.isPartialMatch ? "yellow" : "red",
      color: "white"
    });
    return table.render();
  });
  return tables.join("\n");
};
var bold = (text) => `\x1B[1m${text}\x1B[22m`;
var createTableForRequest = (request) => {
  if (!ttyTable) {
    return [
      `  ${bold("Url".padEnd(8))} ${decodeURIComponent(sortedUrl(request.url()))}`,
      `  ${"Method".padEnd(8)} ${request.method()}`,
      `  ${"Body".padEnd(8)} ${decodeURIComponent(toSortedQueryString(request.body()))}`,
      `  ${"Headers".padEnd(8)} ${decodeURIComponent(toSortedQueryString(request.headers()))}`
    ].join("\n");
  }
  const header = [
    { value: "Outbound Request", width: 20, color: "white", align: "left" },
    { value: "Value", width: 70, color: "white", align: "left" }
  ];
  const rows = [
    [bold("Url"), decodeURIComponent(sortedUrl(request.url()))],
    ["Method", request.method()],
    ["Body", decodeURIComponent(toSortedQueryString(request.body()))],
    ["Headers", decodeURIComponent(toSortedQueryString(request.headers()))]
  ];
  const table = ttyTable(header, rows, {
    borderStyle: "solid",
    borderColor: "blue",
    color: "white"
  });
  return table.render();
};
var createColorExplanationTable = () => {
  const text1 = `${colors.green("\u25CF")} Exact match
${colors.yellow("\u25CF")} Partial match (When URL and METHOD match)
${colors.red("\u25CF")} Not matching`;
  const text2 = `${colors.green("Mock value matches the request value")}
${colors.bgGreen("Present in mock but not in request")}
${colors.bgRed("Present in request but not in mock")}`;
  if (!ttyTable) {
    return `${text1}

${text2}`;
  }
  const header = [
    { value: "Colors Explanation", width: 50, headerColor: "white", align: "left" },
    { value: "Diff Explanation", width: 50, headerColor: "white", align: "left" }
  ];
  const rows = [[text1, text2]];
  const table = ttyTable(header, rows, {
    borderStyle: "solid",
    borderColor: "blue",
    color: "white"
  });
  return table.render();
};
var getDebugPrintMessage = (request, matches) => {
  const requestTable = createTableForRequest(request);
  const message = `
------------------------------------------------------------------------------------------------------------------
Mappersmith matches a mock to an outgoing request by comparing the request's URL, METHOD, BODY and HEADERS.

URL:     The URL is sorted and normalized before comparison.
BODY:    The BODY is sorted and normalized in the form of a query-string before comparison.
HEADERS: The headers of the outgoing request is stripped of any headers that are not present in the
         mock definition headers and then sorted and normalized before comparison.
METHOD:  The method of the outgoing request is compared as is.
------------------------------------------------------------------------------------------------------------------

Request:
${requestTable}

Mock definitions installed:

${createColorExplanationTable()}

${matches.length ? createTableForMocks(matches) : "NO MOCKS INSTALLED"}
${!ttyTable || !Diff ? '\nNOTE: Install optional peer dependencies for richer output: "diff" (character-level diffs) and "tty-table" (formatted tables). See https://github.com/tulios/mappersmith#enhanced-debugging' : ""}
`;
  return message;
};
var noMocksInstalledMessage = (request) => {
  const debugInfo = getDebugPrintMessage(request, []);
  return `

${debugInfo}

${colors.red("[Mappersmith Test] There are no mocks installed, please refer to the documentation: https://github.com/tulios/mappersmith?tab=readme-ov-file#-testing-mappersmith")}
`;
};
var noMatchFoundMessage = (request, matches) => {
  const debugInfo = getDebugPrintMessage(request, matches);
  return `

${debugInfo}

${colors.red("[Mappersmith Test] No match found, check your mock definitions, debug information available above.")}
`;
};
var partialMatchMessage = (request, matches) => {
  const debugInfo = getDebugPrintMessage(request, matches);
  return `

${debugInfo}

${colors.yellow(`[Mappersmith Test] No exact match found but a partial match was found, check your mock definitions, debug information available above.`)}
`;
};
var lookupResponse = (store) => (request) => {
  if (!store.length) {
    throw new Error(noMocksInstalledMessage(request));
  }
  const mocks = store.map((mockResource) => {
    const mockRequest = mockResource.toMockRequest();
    const requestMatching = mockRequest.getRequestMatching(request);
    return {
      requestMatching,
      mockRequest,
      mockResource
    };
  });
  const exactMatch = mocks.filter(({ requestMatching }) => requestMatching.isExactMatch).pop();
  if (exactMatch) {
    return exactMatch.mockRequest.call(request);
  }
  const partialMatch = mocks.filter(({ requestMatching }) => requestMatching.isPartialMatch).pop();
  if (partialMatch) {
    throw new Error(
      partialMatchMessage(
        request,
        mocks.map(({ requestMatching }) => requestMatching)
      )
    );
  }
  throw new Error(
    noMatchFoundMessage(
      request,
      mocks.map(({ requestMatching }) => requestMatching)
    )
  );
};
export {
  diffString,
  getDebugPrintMessage,
  lookupResponse,
  noMatchFoundMessage,
  partialMatchMessage
};
